source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -0,0 +1,426 @@
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * Modifies the state of the filter by adding response headers to the decoder_callbacks
+ */
+Http::FilterHeadersStatus OAuth2Filter::signOutUser(const Http::RequestHeaderMap& headers) {
+  Http::ResponseHeaderMapPtr response_headers{Http::createHeaderMap<Http::ResponseHeaderMapImpl>(
+      {{Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::Found))}})};
+
+  const std::string new_path =
+      absl::StrCat(headers.ForwardedProto()->value().getStringView(), "://", host_, "/");
+  response_headers->addReference(Http::Headers::get().SetCookie, SignoutCookieValue);
+  response_headers->addReference(Http::Headers::get().SetCookie, SignoutBearerTokenValue);
+  response_headers->setLocation(new_path);
+  decoder_callbacks_->encodeHeaders(std::move(response_headers), true);
+
+  return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+}
+
+void OAuth2Filter::onGetAccessTokenSuccess(const std::string& access_code,
+                                           std::chrono::seconds expires_in) {
+  access_token_ = access_code;
+
+  const auto new_epoch = time_source_.systemTime() + expires_in;
+  new_expires_ = std::to_string(
+      std::chrono::duration_cast<std::chrono::seconds>(new_epoch.time_since_epoch()).count());
+
+  finishFlow();
+}
+
+void OAuth2Filter::finishFlow() {
+
+  // We have fully completed the entire OAuth flow, whether through Authorization header or from
+  // user redirection to the auth server.
+  if (found_bearer_token_) {
+    setXForwardedOauthHeaders(*request_headers_, access_token_);
+    config_->stats().oauth_success_.inc();
+    decoder_callbacks_->continueDecoding();
+    return;
+  }
+
+  std::string token_payload;
+  if (config_->forwardBearerToken()) {
+    token_payload = absl::StrCat(host_, new_expires_, access_token_);
+  } else {
+    token_payload = absl::StrCat(host_, new_expires_);
+  }
+
+  auto& crypto_util = Envoy::Common::Crypto::UtilitySingleton::get();
+
+  auto token_secret = config_->tokenSecret();
+  std::vector<uint8_t> token_secret_vec(token_secret.begin(), token_secret.end());
+  const std::string pre_encoded_token =
+      Hex::encode(crypto_util.getSha256Hmac(token_secret_vec, token_payload));
+  std::string encoded_token;
+  absl::Base64Escape(pre_encoded_token, &encoded_token);
+
+  // We use HTTP Only cookies for the HMAC and Expiry.
+  const std::string cookie_tail = fmt::format(CookieTailFormatString, new_expires_);
+  const std::string cookie_tail_http_only =
+      fmt::format(CookieTailHttpOnlyFormatString, new_expires_);
+
+  /**
+   * At this point we have all of the pieces needed to authenticate a user that did not originally
+   * have a bearer access token. Now, we construct a redirect request to return the user to their
+   * previous state and additionally set the OAuth cookies in browser.
+   * The redirection should result in successfully passing this filter.
+   */
```

Comment:
This was probably my bad, but this comment should probably be `//`-style for consistency